src/bp/ui-admin/src/telemetry/client.ts

```diff
@@ -0,0 +1,143 @@
+import axios from 'axios'
+import { lang } from 'botpress/shared'
+import _ from 'lodash'
+import moment from 'moment'
+import ms from 'ms'
+import uuid from 'uuid'
+
+import store from '../store'
+
+interface EventPackageInfoType {
+  [key: string]: {
+    locked: boolean
+    timeout: string
+    getPackage: Function
+  }
+}
+
+export const axiosConfig = {
+  baseURL: window.TELEMETRY_URL,
+  headers: {
+    withCredentials: false
+  }
+}
+
+const dataSchemaVersion = '1.0.0'
+
+const telemetrySchemaVersion = '1.0.0'
+
+const storeInfos: Function[] = []
+
+const eventPackageInfo: EventPackageInfoType = {}
+
+const toggleLock = (eventName: string) => {
+  if (_.has(eventPackageInfo, eventName)) {
+    eventPackageInfo[eventName].locked = !eventPackageInfo[eventName].locked
+  }
+}
+
+const getEventLockTimeout = (event: string) => {
+  const item = window.localStorage.getItem(event)
+  if (item !== null) {
+    const timeout = parseInt(item) - new Date().getTime()
+    if (timeout > 0) {
+      return timeout
+    }
+    window.localStorage.removeItem(event)
+  }
+  return 0
+}
+
+const addEventLockTimeout = (event: string, timeout: number) => {
+  setTimeout(() => toggleLock(event), timeout)
+  window.localStorage.setItem(event, (timeout + moment().valueOf()).toString())
+}
+
+const checkStoreInfoReceived = () => !storeInfos.find(info => info() === undefined)
+
+export const trackReduxStoreInfo = (pathInStore: string) => {
+  storeInfos.push(() => {
+    return _.get(store.getState(), pathInStore)
+  })
+}
+
+export const addTelemetryEvent = (name: string, timeout: string, getPackage: Function) => {
+  eventPackageInfo[name] = {
+    locked: getEventLockTimeout(name) >= 0,
+    timeout,
+    getPackage
+  }
+}
+
+const checkTelemetry = async () => {
+  for (const eventName in eventPackageInfo) {
+    if (!eventPackageInfo[eventName].locked) {
+      toggleLock(eventName)
+
+      try {
+        const pkg = eventPackageInfo[eventName].getPackage()
+
+        if (typeof pkg === 'object') {
+          await sendTelemetryEvent(pkg, eventName)
+        } else {
+          console.error('The package received was incorrect', pkg)
+        }
+      } catch (err) {
+        console.error(`Could not send the telemetry package to the storage server`, err)
+      }
+    }
+  }
+}
+
+export const startTelemetry = () => {
+  trackReduxStoreInfo('user.profile.email')
+
+  trackReduxStoreInfo('version.currentVersion')
+
+  trackReduxStoreInfo('license.licensing.isPro')
+
+  addTelemetryEvent('ui_language', '8h', () => {
+    return {
+      user: {
+        email: store.getState().user.profile.email,
+        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone
+      },
```

Comment:
Please remove all that logic (trackRedux, addTelemetry), there's not enough stuff to track to implement a complicated mechanism. Keep it simple

Reply:
trackRedux can be easily removed and still work the same. On the other hand, being able to easily add new data to analyse was a caracteristic that our code needed to have for the future. Because my part was a bit smaller than @chsir17, I made my code a bit more flexible at the same time. It is what the function addTelemetry is supposed to do

Reply:
I don't see how it would be more complicated to do something simple like
```js
const metrics = ['user.profile.email','version.currentVersion','license.licensing.isPro' ]
  const state = store.getState()
  const values = metrics.map(x => _.get(state, x))
```

We're not gonna have 100 stuff to track, we need to keep it readable and simple to understand. Don't over-engineer it, if you have 3 informations to track right now, build it to track 3. When we need to add more we'll optimize if needed

Reply:
Fixed in : 
52d479c?